{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "f84ca06c-7681-4760-b9e9-48f5eb69f352",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "\t\t\t<script type=\"text/javascript\">\n",
       "\t\t\t<!--\n",
       "\t\t\t\t\n",
       "\t\t\t{\n",
       "\t\t\t\tvar element = document.getElementById('NetworKit_script');\n",
       "\t\t\t\tif (element) {\n",
       "\t\t\t\t\telement.parentNode.removeChild(element);\n",
       "\t\t\t\t}\n",
       "\t\t\t\telement = document.createElement('script');\n",
       "\t\t\t\telement.type = 'text/javascript';\n",
       "\t\t\t\telement.innerHTML = 'function NetworKit_pageEmbed(id) { var i, j; var elements; elements = document.getElementById(id).getElementsByClassName(\"Plot\"); for (i=0; i<elements.length; i++) { elements[i].id = id + \"_Plot_\" + i; var data = elements[i].getAttribute(\"data-image\").split(\"|\"); elements[i].removeAttribute(\"data-image\"); var content = \"<div class=\\\\\"Image\\\\\" id=\\\\\"\" + elements[i].id + \"_Image\\\\\" />\"; elements[i].innerHTML = content; elements[i].setAttribute(\"data-image-index\", 0); elements[i].setAttribute(\"data-image-length\", data.length); for (j=0; j<data.length; j++) { elements[i].setAttribute(\"data-image-\" + j, data[j]); } NetworKit_plotUpdate(elements[i]); elements[i].onclick = function (e) { NetworKit_overlayShow((e.target) ? e.target : e.srcElement); } } elements = document.getElementById(id).getElementsByClassName(\"HeatCell\"); for (i=0; i<elements.length; i++) { var data = parseFloat(elements[i].getAttribute(\"data-heat\")); var color = \"#00FF00\"; if (data <= 1 && data > 0) { color = \"hsla(0, 100%, 75%, \" + (data) + \")\"; } else if (data <= 0 && data >= -1) { color = \"hsla(240, 100%, 75%, \" + (-data) + \")\"; } elements[i].style.backgroundColor = color; } elements = document.getElementById(id).getElementsByClassName(\"Details\"); for (i=0; i<elements.length; i++) { elements[i].setAttribute(\"data-title\", \"-\"); NetworKit_toggleDetails(elements[i]); elements[i].onclick = function (e) { NetworKit_toggleDetails((e.target) ? e.target : e.srcElement); } } elements = document.getElementById(id).getElementsByClassName(\"MathValue\"); for (i=elements.length-1; i>=0; i--) { value = elements[i].innerHTML.trim(); if (value === \"nan\") { elements[i].parentNode.innerHTML = \"\" } } elements = document.getElementById(id).getElementsByClassName(\"SubCategory\"); for (i=elements.length-1; i>=0; i--) { value = elements[i].innerHTML.trim(); if (value === \"\") { elements[i].parentNode.removeChild(elements[i]) } } elements = document.getElementById(id).getElementsByClassName(\"Category\"); for (i=elements.length-1; i>=0; i--) { value = elements[i].innerHTML.trim(); if (value === \"\") { elements[i].parentNode.removeChild(elements[i]) } } var isFirefox = false; try { isFirefox = typeof InstallTrigger !== \"undefined\"; } catch (e) {} if (!isFirefox) { alert(\"Currently the function\\'s output is only fully supported by Firefox.\"); } } function NetworKit_plotUpdate(source) { var index = source.getAttribute(\"data-image-index\"); var data = source.getAttribute(\"data-image-\" + index); var image = document.getElementById(source.id + \"_Image\"); image.style.backgroundImage = \"url(\" + data + \")\"; } function NetworKit_showElement(id, show) { var element = document.getElementById(id); element.style.display = (show) ? \"block\" : \"none\"; } function NetworKit_overlayShow(source) { NetworKit_overlayUpdate(source); NetworKit_showElement(\"NetworKit_Overlay\", true); } function NetworKit_overlayUpdate(source) { document.getElementById(\"NetworKit_Overlay_Title\").innerHTML = source.title; var index = source.getAttribute(\"data-image-index\"); var data = source.getAttribute(\"data-image-\" + index); var image = document.getElementById(\"NetworKit_Overlay_Image\"); image.setAttribute(\"data-id\", source.id); image.style.backgroundImage = \"url(\" + data + \")\"; var link = document.getElementById(\"NetworKit_Overlay_Toolbar_Bottom_Save\"); link.href = data; link.download = source.title + \".svg\"; } function NetworKit_overlayImageShift(delta) { var image = document.getElementById(\"NetworKit_Overlay_Image\"); var source = document.getElementById(image.getAttribute(\"data-id\")); var index = parseInt(source.getAttribute(\"data-image-index\")); var length = parseInt(source.getAttribute(\"data-image-length\")); var index = (index+delta) % length; if (index < 0) { index = length + index; } source.setAttribute(\"data-image-index\", index); NetworKit_overlayUpdate(source); } function NetworKit_toggleDetails(source) { var childs = source.children; var show = false; if (source.getAttribute(\"data-title\") == \"-\") { source.setAttribute(\"data-title\", \"+\"); show = false; } else { source.setAttribute(\"data-title\", \"-\"); show = true; } for (i=0; i<childs.length; i++) { if (show) { childs[i].style.display = \"block\"; } else { childs[i].style.display = \"none\"; } } }';\n",
       "\t\t\t\telement.setAttribute('id', 'NetworKit_script');\n",
       "\t\t\t\tdocument.head.appendChild(element);\n",
       "\t\t\t}\n",
       "\t\t\n",
       "\t\t\t\t\n",
       "\t\t\t{\n",
       "\t\t\t\tvar element = document.getElementById('NetworKit_style');\n",
       "\t\t\t\tif (element) {\n",
       "\t\t\t\t\telement.parentNode.removeChild(element);\n",
       "\t\t\t\t}\n",
       "\t\t\t\telement = document.createElement('style');\n",
       "\t\t\t\telement.type = 'text/css';\n",
       "\t\t\t\telement.innerHTML = '.NetworKit_Page { font-family: Arial, Helvetica, sans-serif; font-size: 14px; } .NetworKit_Page .Value:before { font-family: Arial, Helvetica, sans-serif; font-size: 1.05em; content: attr(data-title) \":\"; margin-left: -2.5em; padding-right: 0.5em; } .NetworKit_Page .Details .Value:before { display: block; } .NetworKit_Page .Value { font-family: monospace; white-space: pre; padding-left: 2.5em; white-space: -moz-pre-wrap !important; white-space: -pre-wrap; white-space: -o-pre-wrap; white-space: pre-wrap; word-wrap: break-word; tab-size: 4; -moz-tab-size: 4; } .NetworKit_Page .Category { clear: both; padding-left: 1em; margin-bottom: 1.5em; } .NetworKit_Page .Category:before { content: attr(data-title); font-size: 1.75em; display: block; margin-left: -0.8em; margin-bottom: 0.5em; } .NetworKit_Page .SubCategory { margin-bottom: 1.5em; padding-left: 1em; } .NetworKit_Page .SubCategory:before { font-size: 1.6em; display: block; margin-left: -0.8em; margin-bottom: 0.5em; } .NetworKit_Page .SubCategory[data-title]:before { content: attr(data-title); } .NetworKit_Page .Block { display: block; } .NetworKit_Page .Block:after { content: \".\"; visibility: hidden; display: block; height: 0; clear: both; } .NetworKit_Page .Block .Thumbnail_Overview, .NetworKit_Page .Block .Thumbnail_ScatterPlot { width: 260px; float: left; } .NetworKit_Page .Block .Thumbnail_Overview img, .NetworKit_Page .Block .Thumbnail_ScatterPlot img { width: 260px; } .NetworKit_Page .Block .Thumbnail_Overview:before, .NetworKit_Page .Block .Thumbnail_ScatterPlot:before { display: block; text-align: center; font-weight: bold; } .NetworKit_Page .Block .Thumbnail_Overview:before { content: attr(data-title); } .NetworKit_Page .HeatCell { font-family: \"Courier New\", Courier, monospace; cursor: pointer; } .NetworKit_Page .HeatCell, .NetworKit_Page .HeatCellName { display: inline; padding: 0.1em; margin-right: 2px; background-color: #FFFFFF } .NetworKit_Page .HeatCellName { margin-left: 0.25em; } .NetworKit_Page .HeatCell:before { content: attr(data-heat); display: inline-block; color: #000000; width: 4em; text-align: center; } .NetworKit_Page .Measure { clear: both; } .NetworKit_Page .Measure .Details { cursor: pointer; } .NetworKit_Page .Measure .Details:before { content: \"[\" attr(data-title) \"]\"; display: block; } .NetworKit_Page .Measure .Details .Value { border-left: 1px dotted black; margin-left: 0.4em; padding-left: 3.5em; pointer-events: none; } .NetworKit_Page .Measure .Details .Spacer:before { content: \".\"; opacity: 0.0; pointer-events: none; } .NetworKit_Page .Measure .Plot { width: 440px; height: 440px; cursor: pointer; float: left; margin-left: -0.9em; margin-right: 20px; } .NetworKit_Page .Measure .Plot .Image { background-repeat: no-repeat; background-position: center center; background-size: contain; height: 100%; pointer-events: none; } .NetworKit_Page .Measure .Stat { width: 500px; float: left; } .NetworKit_Page .Measure .Stat .Group { padding-left: 1.25em; margin-bottom: 0.75em; } .NetworKit_Page .Measure .Stat .Group .Title { font-size: 1.1em; display: block; margin-bottom: 0.3em; margin-left: -0.75em; border-right-style: dotted; border-right-width: 1px; border-bottom-style: dotted; border-bottom-width: 1px; background-color: #D0D0D0; padding-left: 0.2em; } .NetworKit_Page .Measure .Stat .Group .List { -webkit-column-count: 3; -moz-column-count: 3; column-count: 3; } .NetworKit_Page .Measure .Stat .Group .List .Entry { position: relative; line-height: 1.75em; } .NetworKit_Page .Measure .Stat .Group .List .Entry[data-tooltip]:before { position: absolute; left: 0; top: -40px; background-color: #808080; color: #ffffff; height: 30px; line-height: 30px; border-radius: 5px; padding: 0 15px; content: attr(data-tooltip); white-space: nowrap; display: none; } .NetworKit_Page .Measure .Stat .Group .List .Entry[data-tooltip]:after { position: absolute; left: 15px; top: -10px; border-top: 7px solid #808080; border-left: 7px solid transparent; border-right: 7px solid transparent; content: \"\"; display: none; } .NetworKit_Page .Measure .Stat .Group .List .Entry[data-tooltip]:hover:after, .NetworKit_Page .Measure .Stat .Group .List .Entry[data-tooltip]:hover:before { display: block; } .NetworKit_Page .Measure .Stat .Group .List .Entry .MathValue { font-family: \"Courier New\", Courier, monospace; } .NetworKit_Page .Measure:after { content: \".\"; visibility: hidden; display: block; height: 0; clear: both; } .NetworKit_Page .PartitionPie { clear: both; } .NetworKit_Page .PartitionPie img { width: 600px; } #NetworKit_Overlay { left: 0px; top: 0px; display: none; position: absolute; width: 100%; height: 100%; background-color: rgba(0,0,0,0.6); z-index: 1000; } #NetworKit_Overlay_Title { position: absolute; color: white; transform: rotate(-90deg); width: 32em; height: 32em; padding-right: 0.5em; padding-top: 0.5em; text-align: right; font-size: 40px; } #NetworKit_Overlay .button { background: white; cursor: pointer; } #NetworKit_Overlay .button:before { size: 13px; display: inline-block; text-align: center; margin-top: 0.5em; margin-bottom: 0.5em; width: 1.5em; height: 1.5em; } #NetworKit_Overlay .icon-close:before { content: \"X\"; } #NetworKit_Overlay .icon-previous:before { content: \"P\"; } #NetworKit_Overlay .icon-next:before { content: \"N\"; } #NetworKit_Overlay .icon-save:before { content: \"S\"; } #NetworKit_Overlay_Toolbar_Top, #NetworKit_Overlay_Toolbar_Bottom { position: absolute; width: 40px; right: 13px; text-align: right; z-index: 1100; } #NetworKit_Overlay_Toolbar_Top { top: 0.5em; } #NetworKit_Overlay_Toolbar_Bottom { Bottom: 0.5em; } #NetworKit_Overlay_ImageContainer { position: absolute; top: 5%; left: 5%; height: 90%; width: 90%; background-repeat: no-repeat; background-position: center center; background-size: contain; } #NetworKit_Overlay_Image { height: 100%; width: 100%; background-repeat: no-repeat; background-position: center center; background-size: contain; }';\n",
       "\t\t\t\telement.setAttribute('id', 'NetworKit_style');\n",
       "\t\t\t\tdocument.head.appendChild(element);\n",
       "\t\t\t}\n",
       "\t\t\n",
       "\t\t\t\t\n",
       "\t\t\t{\n",
       "\t\t\t\tvar element = document.getElementById('NetworKit_Overlay');\n",
       "\t\t\t\tif (element) {\n",
       "\t\t\t\t\telement.parentNode.removeChild(element);\n",
       "\t\t\t\t}\n",
       "\t\t\t\telement = document.createElement('div');\n",
       "\t\t\t\telement.innerHTML = '<div id=\"NetworKit_Overlay_Toolbar_Top\"><div class=\"button icon-close\" id=\"NetworKit_Overlay_Close\" /></div><div id=\"NetworKit_Overlay_Title\" /> <div id=\"NetworKit_Overlay_ImageContainer\"> <div id=\"NetworKit_Overlay_Image\" /> </div> <div id=\"NetworKit_Overlay_Toolbar_Bottom\"> <div class=\"button icon-previous\" onclick=\"NetworKit_overlayImageShift(-1)\" /> <div class=\"button icon-next\" onclick=\"NetworKit_overlayImageShift(1)\" /> <a id=\"NetworKit_Overlay_Toolbar_Bottom_Save\"><div class=\"button icon-save\" /></a> </div>';\n",
       "\t\t\t\telement.setAttribute('id', 'NetworKit_Overlay');\n",
       "\t\t\t\tdocument.body.appendChild(element);\n",
       "\t\t\t\tdocument.getElementById('NetworKit_Overlay_Close').onclick = function (e) {\n",
       "\t\t\t\t\tdocument.getElementById('NetworKit_Overlay').style.display = 'none';\n",
       "\t\t\t\t}\n",
       "\t\t\t}\n",
       "\t\t\n",
       "\t\t\t-->\n",
       "\t\t\t</script>\n",
       "\t\t"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import networkx as nx\n",
    "import random\n",
    "import itertools\n",
    "import gzip\n",
    "import pyranges as pr\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import matplotlib.cm as cm\n",
    "import matplotlib.colors as colors\n",
    "\n",
    "from GraphRicciCurvature.OllivierRicci import OllivierRicci"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "748eeda8-3fd3-42ef-b048-e35d8134db47",
   "metadata": {},
   "source": [
    "# Load data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "c2f69102-7960-4683-a59f-27d5e16e2cc0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "df.shape=(1000000, 10)\n",
      "CPU times: user 869 ms, sys: 279 ms, total: 1.15 s\n",
      "Wall time: 1.16 s\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>readID</th>\n",
       "      <th>chrom1</th>\n",
       "      <th>pos1</th>\n",
       "      <th>chrom2</th>\n",
       "      <th>pos2</th>\n",
       "      <th>strand1</th>\n",
       "      <th>strand2</th>\n",
       "      <th>pair_type</th>\n",
       "      <th>mapq1</th>\n",
       "      <th>mapq2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>81236149</td>\n",
       "      <td>chr1</td>\n",
       "      <td>3000037</td>\n",
       "      <td>chr1</td>\n",
       "      <td>3000363</td>\n",
       "      <td>+</td>\n",
       "      <td>-</td>\n",
       "      <td>UU</td>\n",
       "      <td>14</td>\n",
       "      <td>52</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>53397258</td>\n",
       "      <td>chr1</td>\n",
       "      <td>3000046</td>\n",
       "      <td>chr1</td>\n",
       "      <td>3000430</td>\n",
       "      <td>+</td>\n",
       "      <td>-</td>\n",
       "      <td>UU</td>\n",
       "      <td>9</td>\n",
       "      <td>60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>31007112</td>\n",
       "      <td>chr1</td>\n",
       "      <td>3000152</td>\n",
       "      <td>chr1</td>\n",
       "      <td>3000336</td>\n",
       "      <td>+</td>\n",
       "      <td>-</td>\n",
       "      <td>UU</td>\n",
       "      <td>24</td>\n",
       "      <td>60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4694079</td>\n",
       "      <td>chr1</td>\n",
       "      <td>3000181</td>\n",
       "      <td>chr1</td>\n",
       "      <td>3000644</td>\n",
       "      <td>+</td>\n",
       "      <td>-</td>\n",
       "      <td>UU</td>\n",
       "      <td>31</td>\n",
       "      <td>60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>175297042</td>\n",
       "      <td>chr1</td>\n",
       "      <td>3000254</td>\n",
       "      <td>chr1</td>\n",
       "      <td>3000563</td>\n",
       "      <td>+</td>\n",
       "      <td>-</td>\n",
       "      <td>UU</td>\n",
       "      <td>52</td>\n",
       "      <td>25</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      readID chrom1     pos1 chrom2     pos2 strand1 strand2 pair_type  mapq1  \\\n",
       "0   81236149   chr1  3000037   chr1  3000363       +       -        UU     14   \n",
       "1   53397258   chr1  3000046   chr1  3000430       +       -        UU      9   \n",
       "2   31007112   chr1  3000152   chr1  3000336       +       -        UU     24   \n",
       "3    4694079   chr1  3000181   chr1  3000644       +       -        UU     31   \n",
       "4  175297042   chr1  3000254   chr1  3000563       +       -        UU     52   \n",
       "\n",
       "   mapq2  \n",
       "0     52  \n",
       "1     60  \n",
       "2     60  \n",
       "3     60  \n",
       "4     25  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%time\n",
    "fpath = \"/nfs/turbo/umms-indikar/shared/projects/poreC/data/f1219_population_hic/4DNFI9H5B7J4.pairs.gz\"\n",
    "nrows = 1000000\n",
    "\n",
    "def load_pairs(fpath, nrows=None, usecols=None):\n",
    "    # discover column names from the '#columns:' line\n",
    "    cols = None\n",
    "    with gzip.open(fpath, \"rt\") as fh:\n",
    "        for line in fh:\n",
    "            if line.startswith(\"#columns\"):\n",
    "                cols = line.split(\":\", 1)[1].strip().split()\n",
    "                break\n",
    "            if not line.startswith(\"#\"):\n",
    "                break\n",
    "    if cols is None:\n",
    "        # sensible default if no header is present\n",
    "        cols = [\"readID\",\"chr1\",\"pos1\",\"chr2\",\"pos2\",\"strand1\",\"strand2\"]\n",
    "\n",
    "    if usecols is not None:\n",
    "        # preserve file order while subsetting\n",
    "        cols = [c for c in cols if c in usecols]\n",
    "\n",
    "    df = pd.read_csv(\n",
    "        fpath,\n",
    "        sep=\"\\t\",\n",
    "        comment=\"#\",\n",
    "        compression=\"gzip\",\n",
    "        header=None,\n",
    "        names=cols,\n",
    "        usecols=cols,\n",
    "        low_memory=False,\n",
    "        nrows=nrows,\n",
    "    )\n",
    "    return df\n",
    "\n",
    "df = load_pairs(fpath, nrows=nrows)\n",
    "print(f\"{df.shape=}\")\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9ee81fcb-d1d3-48b9-b562-b7750010db8c",
   "metadata": {},
   "source": [
    "# Load genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "c5dda403-a997-46bc-9d54-62fbebbe1980",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene stable ID</th>\n",
       "      <th>Gene stable ID version</th>\n",
       "      <th>Transcript stable ID</th>\n",
       "      <th>Transcript stable ID version</th>\n",
       "      <th>Gene start (bp)</th>\n",
       "      <th>Gene end (bp)</th>\n",
       "      <th>Transcript start (bp)</th>\n",
       "      <th>Transcript end (bp)</th>\n",
       "      <th>Transcription start site (TSS)</th>\n",
       "      <th>Gene name</th>\n",
       "      <th>Gene % GC content</th>\n",
       "      <th>Gene type</th>\n",
       "      <th>Chromosome/scaffold name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSMUSG00000064336</td>\n",
       "      <td>ENSMUSG00000064336.1</td>\n",
       "      <td>ENSMUST00000082387</td>\n",
       "      <td>ENSMUST00000082387.1</td>\n",
       "      <td>1</td>\n",
       "      <td>68</td>\n",
       "      <td>1</td>\n",
       "      <td>68</td>\n",
       "      <td>1</td>\n",
       "      <td>mt-Tf</td>\n",
       "      <td>30.88</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>MT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSMUSG00000064337</td>\n",
       "      <td>ENSMUSG00000064337.1</td>\n",
       "      <td>ENSMUST00000082388</td>\n",
       "      <td>ENSMUST00000082388.1</td>\n",
       "      <td>70</td>\n",
       "      <td>1024</td>\n",
       "      <td>70</td>\n",
       "      <td>1024</td>\n",
       "      <td>70</td>\n",
       "      <td>mt-Rnr1</td>\n",
       "      <td>35.81</td>\n",
       "      <td>Mt_rRNA</td>\n",
       "      <td>MT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSMUSG00000064338</td>\n",
       "      <td>ENSMUSG00000064338.1</td>\n",
       "      <td>ENSMUST00000082389</td>\n",
       "      <td>ENSMUST00000082389.1</td>\n",
       "      <td>1025</td>\n",
       "      <td>1093</td>\n",
       "      <td>1025</td>\n",
       "      <td>1093</td>\n",
       "      <td>1025</td>\n",
       "      <td>mt-Tv</td>\n",
       "      <td>39.13</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>MT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSMUSG00000064339</td>\n",
       "      <td>ENSMUSG00000064339.1</td>\n",
       "      <td>ENSMUST00000082390</td>\n",
       "      <td>ENSMUST00000082390.1</td>\n",
       "      <td>1094</td>\n",
       "      <td>2675</td>\n",
       "      <td>1094</td>\n",
       "      <td>2675</td>\n",
       "      <td>1094</td>\n",
       "      <td>mt-Rnr2</td>\n",
       "      <td>35.40</td>\n",
       "      <td>Mt_rRNA</td>\n",
       "      <td>MT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSMUSG00000064340</td>\n",
       "      <td>ENSMUSG00000064340.1</td>\n",
       "      <td>ENSMUST00000082391</td>\n",
       "      <td>ENSMUST00000082391.1</td>\n",
       "      <td>2676</td>\n",
       "      <td>2750</td>\n",
       "      <td>2676</td>\n",
       "      <td>2750</td>\n",
       "      <td>2676</td>\n",
       "      <td>mt-Tl1</td>\n",
       "      <td>44.00</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>MT</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       Gene stable ID Gene stable ID version Transcript stable ID  \\\n",
       "0  ENSMUSG00000064336   ENSMUSG00000064336.1   ENSMUST00000082387   \n",
       "1  ENSMUSG00000064337   ENSMUSG00000064337.1   ENSMUST00000082388   \n",
       "2  ENSMUSG00000064338   ENSMUSG00000064338.1   ENSMUST00000082389   \n",
       "3  ENSMUSG00000064339   ENSMUSG00000064339.1   ENSMUST00000082390   \n",
       "4  ENSMUSG00000064340   ENSMUSG00000064340.1   ENSMUST00000082391   \n",
       "\n",
       "  Transcript stable ID version  Gene start (bp)  Gene end (bp)  \\\n",
       "0         ENSMUST00000082387.1                1             68   \n",
       "1         ENSMUST00000082388.1               70           1024   \n",
       "2         ENSMUST00000082389.1             1025           1093   \n",
       "3         ENSMUST00000082390.1             1094           2675   \n",
       "4         ENSMUST00000082391.1             2676           2750   \n",
       "\n",
       "   Transcript start (bp)  Transcript end (bp)  Transcription start site (TSS)  \\\n",
       "0                      1                   68                               1   \n",
       "1                     70                 1024                              70   \n",
       "2                   1025                 1093                            1025   \n",
       "3                   1094                 2675                            1094   \n",
       "4                   2676                 2750                            2676   \n",
       "\n",
       "  Gene name  Gene % GC content Gene type Chromosome/scaffold name  \n",
       "0     mt-Tf              30.88   Mt_tRNA                       MT  \n",
       "1   mt-Rnr1              35.81   Mt_rRNA                       MT  \n",
       "2     mt-Tv              39.13   Mt_tRNA                       MT  \n",
       "3   mt-Rnr2              35.40   Mt_rRNA                       MT  \n",
       "4    mt-Tl1              44.00   Mt_tRNA                       MT  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fpath = \"../resources/mouse_genes.txt.gz\"\n",
    "\n",
    "gdf = pd.read_csv(fpath, sep='\\t', low_memory=False)\n",
    "gdf.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "811f3187-cc66-41ea-bd14-541b01d11ef2",
   "metadata": {},
   "source": [
    "# Filter Genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "163855a2-1f8e-4f37-ab19-e8cb88777202",
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "''",
     "output_type": "error",
     "traceback": [
      "\u001b[31m---------------------------------------------------------------------------\u001b[39m",
      "\u001b[31mKeyError\u001b[39m                                  Traceback (most recent call last)",
      "\u001b[36mFile \u001b[39m\u001b[32m/nfs/turbo/umms-indikar/Cooper/conda_envs/rapids/lib/python3.12/site-packages/pandas/core/indexes/base.py:3805\u001b[39m, in \u001b[36mIndex.get_loc\u001b[39m\u001b[34m(self, key)\u001b[39m\n\u001b[32m   3804\u001b[39m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[32m-> \u001b[39m\u001b[32m3805\u001b[39m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43m_engine\u001b[49m\u001b[43m.\u001b[49m\u001b[43mget_loc\u001b[49m\u001b[43m(\u001b[49m\u001b[43mcasted_key\u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m   3806\u001b[39m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m err:\n",
      "\u001b[36mFile \u001b[39m\u001b[32mindex.pyx:167\u001b[39m, in \u001b[36mpandas._libs.index.IndexEngine.get_loc\u001b[39m\u001b[34m()\u001b[39m\n",
      "\u001b[36mFile \u001b[39m\u001b[32mindex.pyx:196\u001b[39m, in \u001b[36mpandas._libs.index.IndexEngine.get_loc\u001b[39m\u001b[34m()\u001b[39m\n",
      "\u001b[36mFile \u001b[39m\u001b[32mpandas/_libs/hashtable_class_helper.pxi:7081\u001b[39m, in \u001b[36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001b[39m\u001b[34m()\u001b[39m\n",
      "\u001b[36mFile \u001b[39m\u001b[32mpandas/_libs/hashtable_class_helper.pxi:7089\u001b[39m, in \u001b[36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001b[39m\u001b[34m()\u001b[39m\n",
      "\u001b[31mKeyError\u001b[39m: ''",
      "\nThe above exception was the direct cause of the following exception:\n",
      "\u001b[31mKeyError\u001b[39m                                  Traceback (most recent call last)",
      "\u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[4]\u001b[39m\u001b[32m, line 1\u001b[39m\n\u001b[32m----> \u001b[39m\u001b[32m1\u001b[39m \u001b[43mgdf\u001b[49m\u001b[43m[\u001b[49m\u001b[33;43m'\u001b[39;49m\u001b[33;43m'\u001b[39;49m\u001b[43m]\u001b[49m\n",
      "\u001b[36mFile \u001b[39m\u001b[32m/nfs/turbo/umms-indikar/Cooper/conda_envs/rapids/lib/python3.12/site-packages/pandas/core/frame.py:4102\u001b[39m, in \u001b[36mDataFrame.__getitem__\u001b[39m\u001b[34m(self, key)\u001b[39m\n\u001b[32m   4100\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m.columns.nlevels > \u001b[32m1\u001b[39m:\n\u001b[32m   4101\u001b[39m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m._getitem_multilevel(key)\n\u001b[32m-> \u001b[39m\u001b[32m4102\u001b[39m indexer = \u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43mcolumns\u001b[49m\u001b[43m.\u001b[49m\u001b[43mget_loc\u001b[49m\u001b[43m(\u001b[49m\u001b[43mkey\u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m   4103\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m is_integer(indexer):\n\u001b[32m   4104\u001b[39m     indexer = [indexer]\n",
      "\u001b[36mFile \u001b[39m\u001b[32m/nfs/turbo/umms-indikar/Cooper/conda_envs/rapids/lib/python3.12/site-packages/pandas/core/indexes/base.py:3812\u001b[39m, in \u001b[36mIndex.get_loc\u001b[39m\u001b[34m(self, key)\u001b[39m\n\u001b[32m   3807\u001b[39m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(casted_key, \u001b[38;5;28mslice\u001b[39m) \u001b[38;5;129;01mor\u001b[39;00m (\n\u001b[32m   3808\u001b[39m         \u001b[38;5;28misinstance\u001b[39m(casted_key, abc.Iterable)\n\u001b[32m   3809\u001b[39m         \u001b[38;5;129;01mand\u001b[39;00m \u001b[38;5;28many\u001b[39m(\u001b[38;5;28misinstance\u001b[39m(x, \u001b[38;5;28mslice\u001b[39m) \u001b[38;5;28;01mfor\u001b[39;00m x \u001b[38;5;129;01min\u001b[39;00m casted_key)\n\u001b[32m   3810\u001b[39m     ):\n\u001b[32m   3811\u001b[39m         \u001b[38;5;28;01mraise\u001b[39;00m InvalidIndexError(key)\n\u001b[32m-> \u001b[39m\u001b[32m3812\u001b[39m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m(key) \u001b[38;5;28;01mfrom\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[34;01merr\u001b[39;00m\n\u001b[32m   3813\u001b[39m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mTypeError\u001b[39;00m:\n\u001b[32m   3814\u001b[39m     \u001b[38;5;66;03m# If we have a listlike key, _check_indexing_error will raise\u001b[39;00m\n\u001b[32m   3815\u001b[39m     \u001b[38;5;66;03m#  InvalidIndexError. Otherwise we fall through and re-raise\u001b[39;00m\n\u001b[32m   3816\u001b[39m     \u001b[38;5;66;03m#  the TypeError.\u001b[39;00m\n\u001b[32m   3817\u001b[39m     \u001b[38;5;28mself\u001b[39m._check_indexing_error(key)\n",
      "\u001b[31mKeyError\u001b[39m: ''"
     ]
    }
   ],
   "source": [
    "gdf['']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "29cafa69-486c-4a58-881e-3dda7586165e",
   "metadata": {},
   "outputs": [],
   "source": [
    "break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fa4c58c6-5ecd-4e2c-9b50-63c09158a5cf",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(gdf[['Gene name', 'Chromosome/scaffold name', 'Gene start (bp)', 'Gene end (bp)']].head().to_string(index=False))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3d8511cb-633a-447a-99e2-87ee2bba450f",
   "metadata": {},
   "source": [
    "# Merge and Filter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fb037845-31de-4962-8002-0283d616da6a",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time\n",
    "def make_gene_ranges(gdf, chrom=\"chr1\"):\n",
    "    \"\"\"Return PyRanges of genes on `chrom` (1-based inclusive -> 0-based half-open).\"\"\"\n",
    "    g = (gdf.rename(columns={\n",
    "            'Gene name':'gene',\n",
    "            'Chromosome/scaffold name':'chrom',\n",
    "            'Gene start (bp)':'start',\n",
    "            'Gene end (bp)':'end'\n",
    "        }).copy())\n",
    "    g['chrom'] = g['chrom'].astype(str).str.replace('^chr','', regex=True).radd('chr')\n",
    "    g = g[g['chrom'] == chrom]\n",
    "    return pr.PyRanges(\n",
    "        g.assign(Start=g['start']-1, End=g['end'])\n",
    "         [['chrom','Start','End','gene']]\n",
    "         .rename(columns={'chrom':'Chromosome'}).drop_duplicates()\n",
    "    )\n",
    "\n",
    "chrom = 'chr1'\n",
    "gr_genes = make_gene_ranges(gdf, chrom=chrom)\n",
    "print(f\"{len(gr_genes)=}\")\n",
    "print(gr_genes.df.head().to_string(index=False))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fff5e16f-9dc8-47d3-aede-fc3ea82915d5",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time\n",
    "def gene_gene_contacts(df, gr_genes, chrom=\"chr1\"):\n",
    "    \"\"\"Return contacts on `chrom` where both ends overlap a gene in `gr_genes`.\"\"\"\n",
    "    c = df[(df['chrom1'] == chrom) & (df['chrom2'] == chrom)].copy()\n",
    "    if c.empty:\n",
    "        return pd.DataFrame(columns=['chrom1','pos1','chrom2','pos2','gene1','gene2'])\n",
    "\n",
    "    p1 = pr.PyRanges(pd.DataFrame({\n",
    "        'Chromosome': c['chrom1'], 'Start': c['pos1']-1, 'End': c['pos1'], 'i': c.index\n",
    "    }))\n",
    "    p2 = pr.PyRanges(pd.DataFrame({\n",
    "        'Chromosome': c['chrom2'], 'Start': c['pos2']-1, 'End': c['pos2'], 'i': c.index\n",
    "    }))\n",
    "\n",
    "    def end_to_gene(p, label):\n",
    "        j = p.join(gr_genes).as_df()\n",
    "        if j.empty:\n",
    "            return pd.DataFrame(columns=['i', label])\n",
    "        return j[['i','gene']].rename(columns={'gene': label})\n",
    "\n",
    "    a1 = end_to_gene(p1, 'gene1')\n",
    "    a2 = end_to_gene(p2, 'gene2')\n",
    "\n",
    "    out = (c.assign(i=c.index)\n",
    "             .merge(a1, on='i', how='inner')\n",
    "             .merge(a2, on='i', how='inner')\n",
    "             .drop(columns='i'))\n",
    "    return out[['chrom1','pos1','chrom2','pos2','gene1','gene2']]\n",
    "\n",
    "pdf = gene_gene_contacts(df, gr_genes, chrom=\"chr1\")\n",
    "print(pdf.head().to_string(index=False))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4c6663a0-024a-4ea8-bda6-96f328c49837",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time\n",
    "X = pdf.groupby(['gene1', 'gene2'])['pos1'].count().reset_index()\n",
    "X_p = X.copy()\n",
    "X_p.columns = ['gene2', 'gene1', 'pos1']\n",
    "X = pd.concat([X, X_p], ignore_index=True)\n",
    "X = pd.pivot_table(\n",
    "    X, \n",
    "    index='gene1',\n",
    "    columns='gene2',\n",
    "    values='pos1',\n",
    "    fill_value=0.0,\n",
    ")\n",
    "\n",
    "print(f\"{X.shape=}\")\n",
    "X"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "321e2559-047c-4710-a767-12a691bfd6c8",
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.heatmap(np.log1p(X))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a9212c88-8b5e-4071-b9c2-da48b8231b9d",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "rapids",
   "language": "python",
   "name": "rapids"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
